### ISSUE TITLE
ggplot exception causes the server-side R session to halt

### ISSUE THREAD

**Issue by wliao229** [RBigData/remoter#50]
I am using a ssh tunnel to connect to the server.

```ssh -4 <server addr> -L 55556:127.0.0.1:55555  'Rscript -e "remoter::server(port=55555)"'```

I notice whenever there is an exception using ggplot, the server R process will halt. For example, I receive such error message from the server's log:

```Error in paste0(ego_team, ego_uid) : object 'ego_uid' not found
Calls: <Anonymous> ... f -> scales_add_defaults -> lapply -> FUN -> paste0
Execution halted
```

This error is related to a wrong variable specified for ggplot. It seems like the exception is not handled by the server.

**Comment by snoweye** [comment]
Please check if the [PR](https://github.com/RBigData/remoter/pull/52) fixes the problem. Thanks.

**Comment by wliao229** [comment]
Thank you. I got this error when install it.

```
> devtools::install_github('snoweye/remoter')
Downloading GitHub repo snoweye/remoter@master
pdflatex not found! Not building PDF manual.
   Warning: unknown option ‘--no-staged-install’
✔  checking for file ‘/tmp/Rtmp8K605u/remotes45bc61fefc55/snoweye-remoter-aef7d8b/DESCRIPTION’ ...
─  preparing ‘remoter’:
✔  checking DESCRIPTION meta-information ...
─  installing the package to build vignettes
         -----------------------------------
─  installing *source* package ‘remoter’ ...
   ** using staged installation
   ** R
   ** inst
   ** byte-compile and prepare package for lazy loading
   ** help
   *** installing help indices
   ** building package indices
   ** installing vignettes
   ** testing if installed package can be loaded from temporary location
   ** testing if installed package can be loaded from final location
   ** testing if installed package keeps a record of temporary installation path
   ERROR: hard-coded installation path: please report to the package maintainer and use ‘--no-staged-install’
─  removing ‘/tmp/Rtmp5dWHDT/Rinst480946e73cef/remoter’
         -----------------------------------
   ERROR: package installation failed
Error in (function (command = NULL, args = character(), error_on_status = TRUE,  : 
  System command error
```

**Comment by snoweye** [comment]
Please try the newest update again. Thanks.
- You may need a newer R (maybe > 3.7) to have `--no-staged-install`.
- There is no `src/` in `remoter`, so I have no idea where `hard-coded installation path` comes from.

**Comment by howff** [comment]
I have tried this but I still get a crash when I use ggplot and mistype a variable name
```
[2021-02-05 19:14:33]: client connected
Error in FUN(X[[i]], ...) : object 'Sepal.Widths' not found
Calls: <Anonymous> ... <Anonymous> -> f -> scales_add_defaults -> lapply -> FUN
Execution halted
```

Bonus question: when this gets fixed will it be possible to get a new version pushed to CRAN ?

**Comment by snoweye** [comment]
Try if [this update](https://github.com/snoweye/remoter/commit/af88b4feb45ff421040edc6168819bb16a0489bd) fixes the problems. @howff

**Comment by howff** [comment]
The change from `if (ret$visible && is.gg.ggplot(ret$value) && is.rpng.open())`
to `if (ret$visible && is.gg.ggplot(ret$value))` seems to be the correct fix for my problem, thank you very much.

**Comment by snoweye** [comment]
Thanks. @wrathematics may send [this](https://github.com/RBigData/remoter/pull/61) to CRAN when he has time.

**Comment by wrathematics** [comment]
There are some strange things going on with one of the windows cran machines (which affect remoter) that I'm keeping an eye on. Will submit an update soon.

**Comment by snoweye** [comment]
[0.4-0 on R release](https://builder.r-hub.io/status/remoter_0.4-0.tar.gz-482ffc40f7c013255d0803995cfc1ac9) and [0.4-1 on R release](https://builder.r-hub.io/status/remoter_0.4-1.tar.gz-511c159694f2be394a9d38e64600195e) look both fine to me. Just send to R winbuilder for three R versions which you will receive the report.

**Comment by wrathematics** [comment]
Referring to [this](https://cran.r-project.org/web/checks/check_results_remoter.html).

**Comment by snoweye** [comment]
Yes, I saw that and test for released R on windows. Did you get the same message from CRAN? If so, I need the log file to fix. If not, then you should be fine. I think it is related to `tests/batch.R`. If the server did not receive `shutdown()` from CRAN. It will hang forever because it is a `system()` call in background. I will add an automatic kill in 5 mins for windows if that is possible. Let me try.

**Comment by snoweye** [comment]
I add a file check to make sure server is running in the `tests/batch.R`. Connect check would not work because `zmq_connect` return 0 when `addr = localhost`.

**Comment by snoweye** [comment]
I add all socket options to [pbdZMQ](https://github.com/snoweye/pbdZMQ/commit/dff14fe5e442efe4c28b7007ec8ab681065443da). The `CONNECT_TIMEOUT` with `SNDTIMEO` and `RCVTIMEO` may help on this, but it needs to rewrite a lot of `server()` and `client()` ... This is somehow related to [this](https://github.com/RBigData/remoter/issues/56), too. @wrathematics
